### DISCUSSION TITLE
Execute js file which uses Worker API

### DISCUSSION THREAD

**Discussion by GabrielePicco** [sebastienros/jint#1539] [Q&A]
**Version used**

v3.0.0

**Describe the bug**

I'm trying to use a lib bundled with rollup, which uses the [Worker](https://developer.mozilla.org/en-US/docs/Web/API/Worker) api

**To Reproduce**

The line causing the issue is:

`var browser$1 = Worker;`

which fails with: `JavaScriptErrorWrapperException: Worker is not defined`

**Expected behavior**

Being able to use the Worker api

**Additional context**

I tried also to add `export { Worker } from "worker_threads";` before bundling with rollup and embedding a worker mock with [rollup-plugin-worker-factory](https://github.com/brandonocasey/rollup-plugin-worker-factory), but without success.

- I'm also experimenting with some polyfill (e.g.: https://github.com/chjj/bthreads), but no luck so far.

**Comment by lahma** [comment]
Jint doesn't implement the Worker API so that's why it won't work.

**Reply by GabrielePicco** [reply]
Thanks @lahma . What I'm trying is to bundle the library with a Worker [polyfill](https://www.npmjs.com/package/worker-polyfill), using rollup. That should theoretically work, right?

The [library](https://github.com/iden3/snarkjs) is quite complex, so I didn't manage to pack everything together, but that's the experiment I'm trying. The library also uses wasm binaries, so I'm not entirely sure it would work even if I manage to polyfill the Worker api.

**Reply by lahma** [reply]
Everything is possible in theory, but if any of the polyfills needs native constructs from say Node like threading, it's a dead-end. WASM will be probably the next roadblock at least like you've said.

**Reply by lahma** [reply]
You can of course implement and register the Worker API to Jint and try to bypass the need for polyfill, but that will be some work.

**Reply by Genteure** [reply]
Some ideas:

1. Implement `Worker` (either with polyfill or real multithreading) and `WebAssembly` (possibly with [wasmtime-dotnet](https://github.com/bytecodealliance/wasmtime-dotnet)) in Jint.
2. Find an alternative library implemented in C#.
3. Explore the possibility of running that library in V8 by using [ClearScript](https://github.com/microsoft/ClearScript), assuming WebAssembly would just work in V8.
4. Run that library out of process, in another nodejs project. You could exchange data between the two process with something like HTTP API or gRPC.

If I were to do this, since I didn't find any equivalent library in C#, I would probably choose option 4.
Running out of process to me looks like would require the least effort, definitely work, and have predictable performance.

**Reply by GabrielePicco** [reply]
Thanks @lahma and @Genteure for the useful insights.
I'm trying to make the library usable for Unity, so unfortunately 2,3,4 I think are not viable options in my use-case. 
I can use Jint in Unity, but not ClearScript. I will probably continue experimenting with option 1.

**Reply by GabrielePicco** [reply]
The worker polyfill doesn't seem to refer threading (it runs on the main thread), so perhaps that would work, even if implementing and registering the Worker API to Jint as @lahma suggested would be much better. I will do some investigation, as I'm not familiar with the Jint codebase. Thanks for the help.
